src/gurt/telemetry.c

```diff
@@ -893,19 +844,115 @@ d_tm_increment_counter(struct d_tm_node_t **metric, char *item, ...)
 			*metric = node;
 	}
 
-	if (node->dtn_type == D_TM_COUNTER) {
-		if (node->dtn_protect)
-			D_MUTEX_LOCK(&node->dtn_lock);
-		node->dtn_metric->dtm_data.value++;
-		if (node->dtn_protect)
-			D_MUTEX_UNLOCK(&node->dtn_lock);
+	if (node->dtn_type != D_TM_COUNTER) {
+		rc = -DER_OP_NOT_PERMITTED;
+		D_ERROR("Failed to set counter [%s] on item not a "
+			"counter.  Operation mismatch: " DF_RC "\n",
+			node->dtn_name, DP_RC(rc));
+		goto failure;
+	}
+
+	if (node->dtn_protect)
+		D_MUTEX_LOCK(&node->dtn_lock);
+	node->dtn_metric->dtm_data.value += val;
+	if (node->dtn_protect)
+		D_MUTEX_UNLOCK(&node->dtn_lock);
+
+	return rc;
+
+failure:
+	return rc;
+}
+
+/**
+ * Increment the given counter
+ *
+ * The counter is specified either by an initialized pointer or by a fully
+ * qualified item name.  If an initialized pointer is provided, the metric is
+ * accessed directly.  Otherwise, a lookup is performed on the path name
+ * provided in order to find the specified item.  If the item cannot be found,
+ * it is created.  If the item is created, the callers pointer is initialized
+ * for this item.  The pointer is used for direct access on subsequent calls for
+ * faster access.
+ *
+ * \param[in,out]	metric	Pointer to the metric
+ * \param[in]		fmt	Format specifier for the name and full path of
+ *				the metric followed by optional args to
+ *				populate the string, printf style.
+ *				The format specifier and optional arguments
+ *				are used only if the pointer to the metric
+ *				is NULL.
+ *
+ * \return			DER_SUCCESS		Success
+ *				-DER_EXCEEDS_PATH_LEN	The full name length is
+ *							too long
+ *				-DER_OP_NOT_PERMITTED	Operation not permitted
+ *							because the \a item is
+ *							not a counter
+ *				-DER_UNINIT		API not initialized
+ *				-DER_INVAL		\a metric and \a item
+ *							are NULL
+ */
+int
+d_tm_increment_counter(struct d_tm_node_t **metric, const char *fmt, ...)
```

Comment:
Since this is effectively  `d_tm_add_to_counter` with val = 1, you could create a helper function that implements all the logic, but accepts the `va_list` as a param, and it could be called by both functions to reduce duplication.

Reply:
If @dpquigl security analysis of va_list says that is allowable as a param, then that's fine.  Otherwise, we would want to keep it as is.

Reply:
Hmm, okay. Keep in mind the helper would be a static function, so I don't think it'd be any more prone to injection than the existing entry points.

Reply:
I don't expect it to be a problem either, but I am leaving room for that possibility given what David shared about the security concerns.

Reply:
Maybe it would be cleaner to have less functions, and simply go with d_tm_increment_counter(struct d_tm_node_t **metric, uint64_t val, const char *fmt, ...) and let the caller just supply the value they want?  Wouldn't that be cleaner?

Reply:
That would be fine with me.